File: packages/semantic-cache/src/discovery.ts
import { hostname } from 'node:os';

import type { Valkey } from './types';
import { SemanticCacheUsageError } from './errors';

export const PROTOCOL_VERSION = 1;

export const REGISTRY_KEY = '__betterdb:caches';
export const PROTOCOL_KEY = '__betterdb:protocol';
export const HEARTBEAT_KEY_PREFIX = '__betterdb:heartbeat:';

export const DEFAULT_HEARTBEAT_INTERVAL_MS = 30_000;
export const HEARTBEAT_TTL_SECONDS = 60;

export const CACHE_TYPE = 'semantic_cache' as const;
export type CacheType = typeof CACHE_TYPE;

export interface DiscoveryOptions {
  enabled?: boolean;
  heartbeatIntervalMs?: number;
  includeCategories?: boolean;
}

export interface MarkerMetadata {
  type: CacheType;
  prefix: string;
  version: string;
  protocol_version: number;
  capabilities: string[];
  stats_key: string;
  started_at: string;
  pid?: number;
  hostname?: string;
  [extra: string]: unknown;
}

export interface BuildSemanticMetadataInput {
  name: string;
  version: string;
  defaultThreshold: number;
  categoryThresholds: Record<string, number>;
  uncertaintyBand: number;
  includeCategories: boolean;
}

export function buildSemanticMetadata(input: BuildSemanticMetadataInput): MarkerMetadata {
  const metadata: MarkerMetadata = {
    type: CACHE_TYPE,
    prefix: input.name,
    version: input.version,
    protocol_version: PROTOCOL_VERSION,
    capabilities: ['invalidate', 'similarity_distribution', 'threshold_adjust'],
    index_name: `${input.name}:idx`,
    stats_key: `${input.name}:__stats`,
    config_key: `${input.name}:__config`,
    default_threshold: input.defaultThreshold,
    uncertainty_band: input.uncertaintyBand,
    started_at: new Date().toISOString(),
    pid: process.pid,
    hostname: hostname(),
  };
  if (input.includeCategories && Object.keys(input.categoryThresholds).length > 0) {
    metadata.category_thresholds = { ...input.categoryThresholds };
  }
  return metadata;
}

export interface DiscoveryLogger {
  warn: (msg: string) => void;
  debug: (msg: string) => void;
}

const noopLogger: DiscoveryLogger = {
  warn: () => {},
  debug: () => {},
};

function errMsg(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

export interface DiscoveryManagerDeps {
  client: Valkey;
  name: string;
  metadata: MarkerMetadata;
  heartbeatIntervalMs?: number;
  logger?: DiscoveryLogger;
  onWriteFailed?: () => void;
}

export class DiscoveryManager {
  private readonly client: Valkey;
  private readonly name: string;
  private readonly metadata: MarkerMetadata;
  private readonly heartbeatIntervalMs: number;
  private readonly heartbeatKey: string;
  private readonly logger: DiscoveryLogger;
  private readonly onWriteFailed: () => void;

  private heartbeatHandle: ReturnType<typeof setInterval> | null = null;

  constructor(deps: DiscoveryManagerDeps) {
    this.client = deps.client;
    this.name = deps.name;
    this.metadata = deps.metadata;
    this.heartbeatIntervalMs = deps.heartbeatIntervalMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS;
    this.heartbeatKey = `${HEARTBEAT_KEY_PREFIX}${deps.name}`;
    this.logger = deps.logger ?? noopLogger;
    this.onWriteFailed = deps.onWriteFailed ?? (() => {});
  }

  async register(): Promise<void> {
    const existingJson = await this.safeHget();
    if (existingJson !== null) {
      this.checkCollision(existingJson);
    }

    await this.safeCall(
      () => this.client.hset(REGISTRY_KEY, this.name, JSON.stringify(this.metadata)),
      'HSET registry',
    );
    await this.safeCall(
      () => this.client.set(PROTOCOL_KEY, String(PROTOCOL_VERSION), 'NX'),
      'SET protocol',
    );

    await this.tickHeartbeat();

    this.startHeartbeat();
  }

  async stop(opts: { deleteHeartbeat: boolean }): Promise<void> {
    if (this.heartbeatHandle) {
      clearInterval(this.heartbeatHandle);
      this.heartbeatHandle = null;
    }
    if (!opts.deleteHeartbeat) {
      return;
    }
    try {
      await this.client.del(this.heartbeatKey);
    } catch (err) {
      this.logger.debug(`discovery: DEL heartbeat failed: ${errMsg(err)}`);
    }
  }

  async tickHeartbeat(): Promise<void> {
    const now = new Date().toISOString();
    try {
      await this.client.set(this.heartbeatKey, now, 'EX', HEARTBEAT_TTL_SECONDS);
    } catch (err) {
      this.logger.debug(`discovery: heartbeat SET failed: ${errMsg(err)}`);
      this.onWriteFailed();
    }
  }

  private startHeartbeat(): void {
    if (this.heartbeatHandle) {
      clearInterval(this.heartbeatHandle);
    }
    const handle = setInterval(() => {
      void this.tickHeartbeat();
    }, this.heartbeatIntervalMs);
    handle.unref?.();
    this.heartbeatHandle = handle;
  }

  private async safeHget(): Promise<string | null> {
    try {
      return await this.client.hget(REGISTRY_KEY, this.name);
    } catch (err) {
      this.logger.warn(`discovery: HGET registry failed: ${errMsg(err)}`);
      this.onWriteFailed();
      return null;
    }
  }

  private async safeCall(fn: () => Promise<unknown>, label: string): Promise<void> {
    try {
      await fn();
    } catch (err) {
      this.logger.warn(`discovery: ${label} failed: ${errMsg(err)}`);
      this.onWriteFailed();
    }
  }

  private checkCollision(existingJson: string): void {
    let parsed: Partial<MarkerMetadata>;
    try {
      parsed = JSON.parse(existingJson) as Partial<MarkerMetadata>;
    } catch {
      return;
    }
    if (parsed.type && parsed.type !== CACHE_TYPE) {
      throw new SemanticCacheUsageError(
        `cache name collision: '${this.name}' is already registered as type '${String(parsed.type)}' on this Valkey instance`,
      );
    }
    if (parsed.version && parsed.version !== this.metadata.version) {
      this.logger.warn(
        `discovery: overwriting marker for '${this.name}' (existing version ${String(parsed.version)}, this version ${this.metadata.version})`,
      );
    }
  }
}

File: packages/semantic-cache-py/betterdb_semantic_cache/discovery.py
from __future__ import annotations

import asyncio
import json
import os
import socket
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable

from .errors import SemanticCacheUsageError

PROTOCOL_VERSION = 1

REGISTRY_KEY = '__betterdb:caches'
PROTOCOL_KEY = '__betterdb:protocol'
HEARTBEAT_KEY_PREFIX = '__betterdb:heartbeat:'

DEFAULT_HEARTBEAT_INTERVAL_S = 30.0
HEARTBEAT_TTL_SECONDS = 60

CACHE_TYPE = 'semantic_cache'

# MarkerMetadata is an open dict — extra keys are allowed.
MarkerMetadata = dict[str, Any]


@dataclass
class DiscoveryOptions:
    enabled: bool = True
    heartbeat_interval_ms: int = 30_000
    include_categories: bool = True


@dataclass
class BuildSemanticMetadataInput:
    name: str
    version: str
    default_threshold: float
    category_thresholds: dict[str, float]
    uncertainty_band: float
    include_categories: bool


def build_semantic_metadata(input: BuildSemanticMetadataInput) -> MarkerMetadata:
    metadata: MarkerMetadata = {
        'type': CACHE_TYPE,
        'prefix': input.name,
        'version': input.version,
        'protocol_version': PROTOCOL_VERSION,
        'capabilities': ['invalidate', 'similarity_distribution', 'threshold_adjust'],
        'index_name': f'{input.name}:idx',
        'stats_key': f'{input.name}:__stats',
        'config_key': f'{input.name}:__config',
        'default_threshold': input.default_threshold,
        'uncertainty_band': input.uncertainty_band,
        'started_at': datetime.now(timezone.utc).isoformat(),
        'pid': os.getpid(),
        'hostname': socket.gethostname(),
    }
    if input.include_categories and input.category_thresholds:
        metadata['category_thresholds'] = dict(input.category_thresholds)
    return metadata


def _err_msg(err: Exception) -> str:
    return str(err)


class DiscoveryManager:
    def __init__(
        self,
        *,
        client: Any,
        name: str,
        build_metadata: Callable[[], MarkerMetadata],
        heartbeat_interval_s: float = DEFAULT_HEARTBEAT_INTERVAL_S,
        logger: Any = None,
        on_write_failed: Callable[[], None] | None = None,
    ) -> None:
        self._client = client
        self._name = name
        self._build_metadata = build_metadata
        self._heartbeat_interval_s = heartbeat_interval_s
        self._heartbeat_key = f'{HEARTBEAT_KEY_PREFIX}{name}'
        self._logger = logger
        self._on_write_failed: Callable[[], None] = on_write_failed or (lambda: None)
        self._heartbeat_task: asyncio.Task[None] | None = None

    def _warn(self, msg: str) -> None:
        if self._logger is not None:
            self._logger.warning(msg)

    def _debug(self, msg: str) -> None:
        if self._logger is not None:
            self._logger.debug(msg)

    async def register(self) -> None:
        existing_json = await self._safe_hget()
        if existing_json is not None:
            self._check_collision(existing_json)

        await self._write_metadata()
        await self._safe_call(
            lambda: self._client.set(PROTOCOL_KEY, str(PROTOCOL_VERSION), 'NX'),
            'SET protocol',
        )

        await self._write_heartbeat()
        self._start_heartbeat()

    async def stop(self, *, delete_heartbeat: bool) -> None:
        if self._heartbeat_task is not None:
            self._heartbeat_task.cancel()
            try:
                await self._heartbeat_task
            except (asyncio.CancelledError, Exception):
                pass
            self._heartbeat_task = None

        if not delete_heartbeat:
            return

        try:
            await self._client.delete(self._heartbeat_key)
        except Exception as err:
            self._debug(f'discovery: DEL heartbeat failed: {_err_msg(err)}')

    async def tick_heartbeat(self) -> None:
        await self._write_heartbeat()
        await self._write_metadata()
        await self._safe_call(
            lambda: self._client.set(PROTOCOL_KEY, str(PROTOCOL_VERSION), 'NX'),
            'SET protocol (heartbeat)',
        )

    def _start_heartbeat(self) -> None:
        async def _loop() -> None:
            try:
                while True:
                    await asyncio.sleep(self._heartbeat_interval_s)
                    await self.tick_heartbeat()
            except asyncio.CancelledError:
                pass

        self._heartbeat_task = asyncio.create_task(_loop())

    async def _write_heartbeat(self) -> None:
        now = datetime.now(timezone.utc).isoformat()
        try:
            await self._client.set(self._heartbeat_key, now, 'EX', HEARTBEAT_TTL_SECONDS)
        except Exception as err:
            self._debug(f'discovery: heartbeat SET failed: {_err_msg(err)}')
            self._on_write_failed()

    async def _write_metadata(self) -> None:
        try:
            payload = json.dumps(self._build_metadata())
        except Exception as err:
            self._warn(f'discovery: metadata serialise failed: {_err_msg(err)}')
            self._on_write_failed()
            return
        await self._safe_call(
            lambda: self._client.hset(REGISTRY_KEY, self._name, payload),
            'HSET registry',
        )

    async def _safe_hget(self) -> str | None:
        try:
            result = await self._client.hget(REGISTRY_KEY, self._name)
            if result is None:
                return None
            return result.decode() if isinstance(result, bytes) else result
        except Exception as err:
            self._warn(f'discovery: HGET registry failed: {_err_msg(err)}')
            self._on_write_failed()
            return None

    async def _safe_call(self, fn: Callable[[], Any], label: str) -> None:
        try:
            await fn()
        except Exception as err:
            self._warn(f'discovery: {label} failed: {_err_msg(err)}')
            self._on_write_failed()

    def _check_collision(self, existing_json: str) -> None:
        try:
            parsed: dict[str, Any] = json.loads(existing_json)
        except Exception:
            return

        existing_type = parsed.get('type')
        if existing_type and existing_type != CACHE_TYPE:
            raise SemanticCacheUsageError(
                f"cache name collision: '{self._name}' is already registered as type "
                f"'{existing_type}' on this Valkey instance"
            )

        new_meta = self._build_metadata()
        existing_version = parsed.get('version')
        new_version = new_meta.get('version')
        if existing_version and existing_version != new_version:
            self._warn(
                f"discovery: overwriting marker for '{self._name}' "
                f"(existing version {existing_version}, this version {new_version})"
            )
